From stern@rowland.harvard.edu  Fri Jan 11 17:14:34 2008
From: Alan Stern <stern@rowland.harvard.edu>
Date: Thu, 10 Jan 2008 11:14:53 -0500 (EST)
Subject: USB: EHCI: add a short delay to the bus_suspend routine
To: Greg KH <greg@kroah.com>
Cc: David Brownell <david-b@pacbell.net>, USB list <linux-usb@vger.kernel.org>
Message-ID: <Pine.LNX.4.44L0.0801101114010.2727-100000@iolanthe.rowland.org>


This patch (as1031) adds a short delay to the bus-suspend routine in
ehci-hcd.  Without it some devices disconnect when they should
suspend.

Signed-off-by: Alan Stern <stern@rowland.harvard.edu>
Acked-by: David Brownell <david-b@pacbell.net>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/host/ehci-hub.c |    4 ++++
 1 file changed, 4 insertions(+)

--- a/drivers/usb/host/ehci-hub.c
+++ b/drivers/usb/host/ehci-hub.c
@@ -172,6 +172,10 @@ static int ehci_bus_suspend (struct usb_
 		}
 	}
 
+	/* Apparently some devices need a >= 1-uframe delay here */
+	if (ehci->bus_suspended)
+		udelay(150);
+
 	/* turn off now-idle HC */
 	ehci_halt (ehci);
 	hcd->state = HC_STATE_SUSPENDED;
